Add dict with key type decoder to higher-order-decoders

Addeuz opened the pull request:
Hello, here is the pull request we talked about in #19 (this PR closes #19).

However I marked it as a draft PR since we have to discuss some things.

First of all I added a new decoder since I didn't know if you wanted me to do that or extend the current dict decoder. But I can extend the current one if you want. And just make the keys argument optional.

Secondly I can't seem to get the default value for `D extends Decoder<unknown>` that we discussed about in the issue to work.
If I put this in the generic parameter: `D extends Decoder<unknown> = Decoder<unknown>`
And we give a type to `K` and not to `D`. And we do this:
```ts
export const SizeValues = ['small', 'medium', 'large'] as const;
type Sizes = typeof SizeValues[number];

const dictDecoder = dictWithTypedKey<Sizes>(boolean, SizeValues);

type DictDecoder = decodeType<typeof dictDecoder>
```
And then hover `DictDecoder`, it says the type is `Map<"small" | "medium" | "large", unknown>`.

If we remove the assignment in the generic parameters and do this:
```ts
const SizeValues = ['small', 'medium', 'large'] as const;
type Sizes = typeof SizeValues[number];

const dictDecoder = dictWithTypedKey(boolean, SizeValues);

type DictDecoder = decodeType<typeof dictDecoder>
```
Then the type `DictDecoder` is inferred correctly to `Map<"small" | "medium" | "large", boolean>`.

I'm fine with merging the decoder without inferring the generic `D` however it would be nice if we could find a way..
Because if we can do it and we do this:
```ts
const ExtraSizeValues = ['xlarge', 'schmedium'] as const;

const SizeValues = ['small', 'medium', 'large'] as const;
type Sizes = typeof SizeValues[number];

const dictDecoder = dictWithTypedKey<Sizes>(boolean, ExtraSizeValues);
```

We get this nice error in code:
```
Argument of type 'readonly ["xlarge", "schmedium"]' is not assignable to parameter of type 'readonly ("small" | "medium" | "large")[]'.
  Type '"xlarge" | "schmedium"' is not assignable to type '"small" | "medium" | "large"'.
    Type '"xlarge"' is not assignable to type '"small" | "medium" | "large"'. Did you mean '"large"'?
```

Commit:
Add decoder to higher-order-decoders

tskj commented:
Thank you for submitting this PR!

At first glance I think this looks very promising. I believe it should be possible to do what you want with the type parameter, at least if nothing else with function overloading. Also, I think combining it with the existing decoder could make sense, if it's possible to do in a backwards compatible way!

Commit:
Overload dict function

Addeuz commented:
In the latest commit I successfully overloaded the function. So now in the two unit tests I only use the `dict` decoder. And types are inferred correctly.

They are however still not inferred correctly when just doing this:
`const sizes = dict<Sizes>(boolean, SizeValues)`
Because using it this way they think I want to use the first overload.. I'm not very good on overloading so if you could take a look at this it would be appreciated.

I can however live with this. I.e. if we want to give a type to the generic parameter `K`, we also have to give a type to `D`. But it is not necessary since the types are inferred from `SizeValues`

tskj commented:
That makes sense! Let me have a look.

tskj commented on src/higher-order-decoders.ts:
```diff
@@ -221,21 +221,37 @@ export const map =
+          if (!isKey(key, keys)) {
+            throw `Key \`${key}\` is not in given keys`;
+          }
+
+          return [key, decode(decoder)(value)] as [K, decodeType<D>];
```
```suggestion
```

I think you can just remove this! It shouldn't really be necessary.

tskj commented:
Okay so after doing some research I don't think this is possible. You are correct that it picks the wrong overload, but for a good reason. According to [this SO answer](https://stackoverflow.com/a/38688143), it is not possible for Typescript to infer only some, but not all, of the type parameters (which is what would be required for your usecase to work as far as I understand). Default parameters like I suggested seems to kind of work, but it's tricking us; what's happening is that we are saying the default parameter is `Decoder<unknown>`, but that is not really what we want - we want Typescript to infer a more specific type, a subtype of `Decoder<unknown>` to be specific, and that is not possible if you also want to supply the other type parameter. So this is mostly a limitation in Typescript itself, that might get resolved at some point.

I believe it might be possible to work around this by "currying" the function, i.e. is split the parameters in two like this: `dict_with_keys(SizeValues)(boolean)`, which would allow you to write any combination of `dict_with_keys<Sizes>(SizeValues)(boolean)` and `dict_with_keys<Sizes>(SizeValues)<Decoder<boolean>>(boolean)` and `dict_with_keys(SizeValues)<Decoder<boolean>>(boolean)`. The reason this would work (I believe) is that you would get two separate parameter lists, which could be inferred or supplied individually.

However, I would suggest you create your own wrapper function like this, that calls into the `dict` you have implemented here. Such a curried function would be too complicated to include in the library, I believe. Also, I question if it's not just as typesafe (and ergonomic) to just not supply any type parameters? Because as you suggest, the types are inferred fine from the actual values. If there is a reason that you want to be explicit about it, I would love to understand it.

I think this is a very neat addition to the library as it works now, so I would very much like to merge this implementation (and I'll happily extend it if Typescript eventually supports partial parameter inference). 

I have suggested a small improvement to the implementation. Also please see if the overloads and default type parameters are still necessary to support the current usecases. Let me know if you're happy with this as it stands and I'll merge it!

Commit:
Fix PR comments

Addeuz commented:
I've fixed the comments in the two latest commits (maybe could've been just one commit if I would've read the whole comment :smile:)
I also changed the order of the generic params to more align with the order of the arguments.

> However, I would suggest you create your own wrapper function like this

That makes sense, however there was no real reason about wanting to give the type as a generic param. More than that was just the first thing that popped into my head as a possible solution. But with the current implementation we still have to write the array which the type is given from. So it wouldn't be necessary for us to give the type explicitly.

And if we need to do it in the future I think a wrapper function would be fine. We already have a decoder that wraps another another decoder. And decodes a lot of stuff in it.

So if you are happy with the implementation it would be superb if you could merge it and publish a new version!

Maybe would need some documentation?

Commit:
Remove overloads and reorder generic param

Addeuz commented:
I've fixed the comments in the two latest commits (maybe could've been just one commit if I would've read the whole comment :smile:)
I also changed the order of the generic params to more align with the order of the arguments.

> However, I would suggest you create your own wrapper function like this

That makes sense, however there was no real reason about wanting to give the type as a generic param. More than that was just the first thing that popped into my head as a possible solution. But with the current implementation we still have to write the array which the type is given from. So it wouldn't be necessary for us to give the type explicitly.

And if we need to do it in the future I think a wrapper function would be fine. We already have a decoder that wraps another another decoder. And decodes a lot of stuff in it.

So if you are happy with the implementation it would be superb if you could merge it and publish a new version!

Maybe would need some documentation?

tskj commented:
Cool, thanks for the good work, I have now merged and released!

Good idea on re-ordering the type parameters, I also added a default value on the `K` parameter, to preserve backwards compatibility (for the case if anyone has `dict<Decoder<boolean>>(...)`. Not really something I think makes that much sense, but we still need to support it since it was possible before. Hopefully this still makes sense with every other usecase we have outlined here, please check the latest commit on master to confirm!
